### PR TITLE
Create 349_IntersectionOfTwoArrays.cpp

### PR DESCRIPTION
**Pull request by andriibugaiov**

### CODE REVIEW

File: cpp/349_IntersectionOfTwoArrays.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// S = O(m + n)
+class Solution {
+public:
+    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
+        // Small optimization to make sure we use less memory
+        // by storing less elements in the unordered_set.
+        if (nums1.size() > nums2.size()) {
+            return intersection(nums2, nums1);
+        }
+        unordered_set<int> unique(nums1.begin(), nums1.end());
```
Comment (Yorkio) [inline cpp/349_IntersectionOfTwoArrays.cpp]: Did not know that we can initialize like this!
Great!

File: cpp/349_IntersectionOfTwoArrays.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            // unordered_set to avoid storing duplicates in
+            // the result.
+            if (unique.find(num) != unique.end()) {
+                result.push_back(num);
+                unique.erase(num);
+            }
+        }
+        return result;
+    }
+};
```
Comment (Yorkio) [inline cpp/349_IntersectionOfTwoArrays.cpp]: This one is good!

File: cpp/349_IntersectionOfTwoArrays.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+// Time/Space analysis:
+// T = O(m + n) - amortized
+// S = O(m + n)
+class Solution {
+public:
+    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
+        // Small optimization to make sure we use less memory
+        // by storing less elements in the unordered_set.
+        if (nums1.size() > nums2.size()) {
```
Comment (Yorkio) [inline cpp/349_IntersectionOfTwoArrays.cpp]: See comment above.

File: 350_IntersectionOfTwoArrays.cpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
+// Definitions:
+// m - the size of the first vector
+// n - the size of the second vector
+
+// Time/Space analysis:
+// T = O(m + n) - amortized
+// S = O(m + n)
+class Solution {
+public:
+    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
```
Comment (Yorkio) [inline 350_IntersectionOfTwoArrays.cpp]: vector<int> intersect ?
Name of the function should be intersection.

File: 350_IntersectionOfTwoArrays.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        vector<int> result;
+        for (int num : nums2) {
+            // If the element from nums2 is in nums1, than it
+            // is in the intersection. We then update the count of the
+            // element in the unordered_map. We remove it from the
+            // unordered_map when the count reached 0 to avoid
+            // storing duplicates in the result.
+            if (unique_count.find(num) != unique_count.end()) {
+                result.push_back(num);
+                if (--unique_count[num] == 0) {
```
Comment (Yorkio) [inline 350_IntersectionOfTwoArrays.cpp]: This is not correct.
I will  break it with test
[1,1]
[1,1]

File: cpp/349_IntersectionOfTwoArrays.cpp (hunk 2 above)
Comment (Yorkio) [inline cpp/349_IntersectionOfTwoArrays.cpp]: But looks like you have not submitted first one.

File: 350_IntersectionOfTwoArrays.cpp (hunk 4 above)
Comment (andriibugaiov) [inline 350_IntersectionOfTwoArrays.cpp]: No. You confused it with problem 349.

File: 350_IntersectionOfTwoArrays.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+public:
+    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
+        // Small optimization to make sure we use less memory
+        // by storing less elements in the unordered_set.
+        if (nums1.size() > nums2.size()) {
+            return intersect(nums2, nums1);
+        }
+        unordered_map<int, int> unique_count;
+        for (int num : nums1) {
+            unique_count[num] += 1;
```
Comment (andriibugaiov) [inline 350_IntersectionOfTwoArrays.cpp]: Yes. If the element is not in map, it is constructed with default constructor (fot int it sets the value to 0).

File: 350_IntersectionOfTwoArrays.cpp (hunk 5 above)
Comment (andriibugaiov) [inline 350_IntersectionOfTwoArrays.cpp]: I think you confused this problem with 349.

File: 350_IntersectionOfTwoArrays.cpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+// n - the size of the second vector
+
+// Time/Space analysis:
+// T = O(m + n) - amortized
+// S = O(m + n)
+class Solution {
+public:
+    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
+        // Small optimization to make sure we use less memory
+        // by storing less elements in the unordered_set.
```
Comment (andriibugaiov) [inline 350_IntersectionOfTwoArrays.cpp]: It does. Still, for Leetcode tests it gives an adventage.

File: cpp/349_IntersectionOfTwoArrays.cpp (hunk 2 above)
Comment (andriibugaiov) [inline cpp/349_IntersectionOfTwoArrays.cpp]: I have combined two problems in a single branch. Hence, single review. The problems are mainly the same.

**Pull request merged**